invokeai/app/services/download/download_base.py

```diff
@@ -0,0 +1,200 @@
+from enum import Enum
+from functools import total_ordering
+from pathlib import Path
+from typing import Callable, List, Optional, Union
+
+from pydantic import BaseModel, Field, PrivateAttr
+from pydantic.networks import AnyHttpUrl
+
+
+class DownloadJobStatus(str, Enum):
+    """State of a download job."""
+
+    WAITING = "waiting"  # not enqueued, will not run
+    RUNNING = "running"  # actively downloading
+    COMPLETED = "completed"  # finished running
+    ERROR = "error"  # terminated with an error message
+
+
+class DownloadJobCancelledException(Exception):
+    """This exception is raised when a download job is cancelled."""
+
+
+class UnknownJobIDException(Exception):
+    """This exception is raised when an invalid job id is referened."""
+
+
+DownloadEventHandler = Callable[
+    [
+        "DownloadJob",
+    ],
+    None,
+]
+
+
+@total_ordering
+class DownloadJob(BaseModel):
+    """Class to monitor and control a model download request."""
+
+    # required variables to be passed in on creation
+    source: AnyHttpUrl = Field(description="Where to download from. Specific types specified in child classes.")
+    dest: Path = Field(description="Destination of downloaded model on local disk; a directory or file path")
+    access_token: Optional[str] = Field(default=None, description="authorization token for protected resources")
+    # automatically assigned on creation
+    id: int = Field(description="Numeric ID of this job", default=-1)  # default id is a sentinel
+    priority: int = Field(default=10, description="Queue priority; lower values are higher priority")
+
+    # set internally during download process
+    status: DownloadJobStatus = Field(default=DownloadJobStatus.WAITING, description="Status of the download")
+    download_path: Optional[Path] = Field(default=None, description="Final location of downloaded file")
+    job_started: Optional[str] = Field(default=None, description="Timestamp for when the download job started")
+    job_ended: Optional[str] = Field(
+        default=None, description="Timestamp for when the download job ende1d (completed or errored)"
+    )
+    bytes: int = Field(default=0, description="Bytes downloaded so far")
+    total_bytes: int = Field(default=0, description="Total file size (bytes)")
+
+    # set when an error occurs
+    error_type: Optional[str] = Field(default=None, description="Name of exception that caused an error")
+    error: Optional[str] = Field(default=None, description="Traceback of the exception that caused an error")
+
+    # internal flag
+    _cancelled: bool = PrivateAttr(default=False)
+
+    # optional event handlers passed in on creation
+    _on_start: Optional[DownloadEventHandler] = PrivateAttr(default=None)
+    _on_progress: Optional[DownloadEventHandler] = PrivateAttr(default=None)
+    _on_complete: Optional[DownloadEventHandler] = PrivateAttr(default=None)
+    _on_error: Optional[DownloadEventHandler] = PrivateAttr(default=None)
```

Comment:
When would we use these instead of the events service?

I suppose in the configuration script, where the events service is unavailable, we'd need to provide handlers like this.

Reply:
I'm planning to have the installer use these callbacks too. It will register downloader `on_progress` callbacks that allow it to monitor the progress of each of the model's component files and issue periodic summed-up progress events.